### DISCUSSION TITLE
How to add new curves?

### DISCUSSION THREAD

**Discussion by m28844** [paulmillr/noble-curves#105] [General]
What files do I need to change to add a new curve to the library? I have to use the brainpool P256R1 curve (and P384, P512) to verify digital signatures. The curve parameters can be found here: https://www.rfc-editor.org/rfc/rfc5639
I copied the files of nistsecP256R1 (p256.*), renamed it to brainpool256r1.* and changed the curve parameters. It compiles, it runs, but the verification fails. What elese needs to be changed to get it wotk?
Alternatively, is there any chance, that the brainpool curves will be added officially to the library any time soon?
Thanks in advance.

**Comment by paulmillr** [comment]
Example: https://github.com/paulmillr/noble-curves#weierstrass-short-weierstrass-curve

Just specifying the parameters should work. You are doing something wrong.

No plans on adding brainpool, unless there would be enough of user requests.

**Reply by tastydev** [reply]
Would also love to see brainpool curves (atleast most common ones like mentioned above) officially be a part of this lib. I see brainpool requests quite often that's why nodejs added native support and i think i also saw a pr on elliptic js package for it.

**Reply by paulmillr** [reply]
Why don't you create a sub-package with brainpool, that depends on noble-curves? If it gains enough adoption, we could incorporate it. We could also point to your project meanwhile, when it's not in curves.

Every line of code that we add needs fuzzing, auditing, support, etc. Quite complex.

**Reply by m28844** [reply]
It's actually not that difficult to adapt the code. I also did it on my own. Just copy a file such as p256.ts and change the parameters of the elliptic curve. These parameters are public. Search the Internet or ask the AI. For brainpoolp256r1 it looks like this:

const Fp = Field(BigInt('0xA9FB57DBA1EEA9BC3E660A909D838D726E3BF623D52620282013481D1F6E5377'));
const CURVE_A = BigInt('0x7D5A0975FC2C3057EEF67530417AFFE7FB8055C126DC5C6CE94A4B44F330B5D9');
const CURVE_B = BigInt('0x26DC5C6CE94A4B44F330B5D9BBD77CBF958416295CF7E1CE6BCCDC18FF8C07B6');

export const brainpoolp256r1 = createCurve({
  a: CURVE_A,
  b: CURVE_B,
  Fp,
  // Curve order, total count of valid points in the field
  n: BigInt('0xA9FB57DBA1EEA9BC3E660A909D838D718C397AA3B561A6F7901E0E82974856A7'),
  // Base (generator) point (x, y)
  Gx: BigInt('0x8BD2AEB9CB7E57CB2C4B482FFC81B7AFB9DE27E1E3BD23C23A4453BD9ACE3262'),
  Gy: BigInt('0x547EF835C3DAC4FD97F8461A14611DC9C27745132DED8E545C1D54C72F046997'),
  h: BigInt(1),
  lowS: false,
} as const);

**Reply by m28844** [reply]
Here are the values for brainpoolp384r1:
const Fp = Field(BigInt('0x8CB91E82A3386D280F5D6F7E50E641DF152F7109ED5456B412B1DA197FB71123ACD3A729901D1A71874700133107EC53'));
const CURVE_A = BigInt('0x7BC382C63D8C150C3C72080ACE05AFA0C2BEA28E4FB22787139165EFBA91F90F8AA5814A503AD4EB04A8C7DD22CE2826');
const CURVE_B = BigInt('0x04A8C7DD22CE28268B39B55416F0447C2FB77DE107DCD2A62E880EA53EEB62D57CB4390295DBC9943AB78696FA504C11');
  n: BigInt('0x8CB91E82A3386D280F5D6F7E50E641DF152F7109ED5456B31F166E6CAC0425A7CF3AB6AF6B7FC3103B883202E9046565'),
  Gx: BigInt('0x1D1C64F068CF45FFA2A63A81B7C13F6B8847A3E77EF14FE3DB7FCAFE0CBD10E8E826E03436D646AAEF87B2E247D4AF1E'),
  Gy: BigInt('0x8ABE1D7520F9C2A45CB1EB8E95CFD55262B70B29FEEC5864E19C054FF99129280E4646217791811142820341263C5315'),
  h: BigInt(1),

And here are the values for brainpoolp512r1:
const Fp = Field(BigInt('0xAADD9DB8DBE9C48B3FD4E6AE33C9FC07CB308DB3B3C9D20ED6639CCA703308717D4D9B009BC66842AECDA12AE6A380E62881FF2F2D82C68528AA6056583A48F3'));
const CURVE_A = BigInt('0x7830A3318B603B89E2327145AC234CC594CBDD8D3DF91610A83441CAEA9863BC2DED5D5AA8253AA10A2EF1C98B9AC8B57F1117A72BF2C7B9E7C1AC4D77FC94CA');
const CURVE_B = BigInt('0x3DF91610A83441CAEA9863BC2DED5D5AA8253AA10A2EF1C98B9AC8B57F1117A72BF2C7B9E7C1AC4D77FC94CADC083E67984050B75EBAE5DD2809BD638016F723');
  n: BigInt('0xAADD9DB8DBE9C48B3FD4E6AE33C9FC07CB308DB3B3C9D20ED6639CCA70330870553E5C414CA92619418661197FAC10471DB1D381085DDADDB58796829CA90069'),
  Gx: BigInt('0x81AEE4BDD82ED9645A21322E9C4C6A9385ED9F70B5D916C1B43B62EEF4D0098EFF3B1F78E2D0D48D50D1687B93B97D5F7C6D5047406A5E688B352209BCB9F822'),
  Gy: BigInt('0x7DDE385D566332ECC0EABFA9CF7822FDF209F70024A57B1AA000C55B881F8111B2DCDE494A5F485E5BCA4BD88A2763AED1CA2B2FA8F0540678CD1E0F3AD80892'),
  h: BigInt(1),

I hope, this helps.

**Reply by stknob** [reply]
Well, let me throw this into the mix: https://github.com/stknob/noble-curves/tree/brainpool-curves
This one includes wycheproof test vectors and (very unofficial) hash-to-curve parameters.

The plan was to completely gut the noble-curves repo and create a brainpool-only add-on package,
at some point.
